esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6118,7 +6104,259 @@ bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &r
+    accessInfo.setFileParts(dfuParts);
+
+    StringBuffer metaInfo, expiryTime;
+    getFileMeta(metaInfo, fileName, fileDesc, udesc, expiryTime, jobId, req);
+    accessInfo.setMetaInfoBlob(metaInfo);
+    accessInfo.setExpiryTime(expiryTime.str());
+
+    if (layoutBin.length() == 0)
+        return;
+
+    if (req.getReturnJsonTypeInfo())
+    {
+        StringBuffer jsonLayout;
+        Owned<IRtlFieldTypeDeserializer> deserializer(createRtlFieldTypeDeserializer());
+        const RtlTypeInfo *typeInfo = deserializer->deserialize(layoutBin);
+        dumpTypeInfo(jsonLayout, typeInfo);
+        if (jsonLayout.length())
+            accessInfo.setRecordTypeInfoJson(jsonLayout.str());
+    }
+    if (req.getReturnBinTypeInfo())
+    {
+        MemoryBuffer binLayout;
+        layoutBin.swapWith(binLayout);
+        if (binLayout.length())
+            accessInfo.setRecordTypeInfoBin(binLayout);
+    }
+}
+
+bool CWsDfuEx::onDFUFileCreate(IEspContext &context, IEspDFUFileCreateRequest &req, IEspDFUFileCreateResponse &resp)
+{
+    try
+    {
+        IConstDFUFileAccessRequestBase &requests = req.getRequestBase();
+        SecAccessFlags accessType = translateToSecAccessFlags(requests.getAccessType());
+        if (SecAccess_None == accessType)
+        {
+            context.setAuthStatus(AUTH_STATUS_NOACCESS);
+            throw MakeStringException(ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUFileCreate - Permission denied.");
+        }
+        context.ensureFeatureAccess(FEATURE_URL, accessType, ECLWATCH_DFU_ACCESS_DENIED, "WsDfu::DFUFileCreate: Permission denied.");
+
+        StringBuffer fileName = req.getName();
+        const char *clusterName = requests.getCluster();
+        const char *recordDefinition = req.getECLRecordDefinition();
+        if (fileName.isEmpty())
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No Name defined.");
+        if (isEmptyString(clusterName))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No Cluster defined.");
+        if (isEmptyString(recordDefinition))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT, "No ECLRecordDefinition defined.");
+
+        MemoryBuffer layoutBin;
+        exportRecordDefinitionBinaryType(recordDefinition, layoutBin);
+
+        StringBuffer userId;
+        Owned<IUserDescriptor> userDesc;
+        context.getUserID(userId);
+        if (!userId.isEmpty())
+        {
+            userDesc.setown(createUserDescriptor());
+            userDesc->set(userId.str(), context.queryPassword(), context.querySignature());
+        }
+
+        fileName.append("_").append(clusterName).append(".tmp");
+        VStringBuffer jobId("Create %s", fileName.str());
+        Owned<IFileDescriptor> fileDesc = createDFUDescriptor(context, fileName.str(), clusterName, jobId.str(), userId.str(), userDesc);
+        Owned<IDistributedFile> file = queryDistributedFileDirectory().createNew(fileDesc);
+        file->setAccessed();
```

Comment:
I'm not sure if IDistributedFile accepts a file part mapping, but if it does, and the user provided one, we'll have to set it around here. @jakesmith

Reply:
As outlined in JIRA (this PR is not targeting correct JIRA at the moment), a list of loctions should be provided by the client (Spark) and the service should create a group. See comments here: https://track.hpccsystems.com/browse/HPCC-20443?focusedCommentId=81520&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-81520